File: frontend/src/lib/msig/render.ts
// Template renderer . browser-side mirror of
// programs/clear-wallet/src/utils/message.rs::render_template.
//
// Given an intent's template string (e.g. `"send {2:10^18} ETH to {1}
// (nonce {0})"`) and its params_data bytes, substitute each `{N}` /
// `{N:10^D}` placeholder with the rendered form of that parameter,
// matching the on-chain builder byte-for-byte.
//
// Critical invariant: the bytes this function emits must equal the
// bytes the on-chain `push_*` calls emit. Any drift (whitespace, digit
// count, hex case) means signatures fail verification. Tests in
// __tests__/render.test.ts pin a golden-vector for each code path.

import bs58 from "bs58";
import {
  ParamType,
  paramOffsetAt,
  type ParamEntry,
} from "@/lib/msig/definition";
import { toHex } from "@/lib/msig/hash";

/// Minimal intent shape the renderer needs. Same fields as
/// `EncodeParamsContext` . intentionally decoupled from the full
/// IntentAccount type so other modules can pass a narrow subset.
export interface RenderContext {
  params: readonly ParamEntry[];
  bytePool: Uint8Array;
  /// The template string stored in `byte_pool[template_offset..+template_len]`.
  template: string;
}

/// Render a template into its final UTF-8 byte form. Non-bracket bytes
/// are copied verbatim; `{N}` placeholders expand via the param at
/// index N; `{N:10^D}` scales a u64 by 10^D and trims trailing zeros.
export function renderTemplate(
  ctx: RenderContext,
  paramsData: Uint8Array
): Uint8Array {
  const encoder = new TextEncoder();
  const tplBytes = encoder.encode(ctx.template);
  const chunks: Uint8Array[] = [];

  let i = 0;
  while (i < tplBytes.length) {
    if (tplBytes[i] === 0x7b /* { */) {
      const end = tplBytes.indexOf(0x7d /* } */, i + 1);
      if (end < 0) {
        throw new Error("renderTemplate: unmatched '{' in template");
      }
      const inner = decodeAscii(tplBytes.subarray(i + 1, end));
      const colonAt = inner.indexOf(":");
      const idxStr = colonAt >= 0 ? inner.slice(0, colonAt) : inner;
      const fmt = colonAt >= 0 ? inner.slice(colonAt + 1) : null;
      const idx = parseDecimalNonNegative(idxStr, "param index");
      chunks.push(renderParam(ctx, paramsData, idx, fmt));
      i = end + 1;
    } else {
      chunks.push(tplBytes.subarray(i, i + 1));
      i += 1;
    }
  }

  return concatBytes(chunks);
}

/// String form of `renderTemplate` . the bytes are valid UTF-8 because
/// every branch writes only ASCII or param-supplied UTF-8 that was
/// originally a validated string.
export function renderTemplateToString(
  ctx: RenderContext,
  paramsData: Uint8Array
): string {
  return new TextDecoder().decode(renderTemplate(ctx, paramsData));
}

// ── parameter rendering ───────────────────────────────────────────────

function renderParam(
  ctx: RenderContext,
  paramsData: Uint8Array,
  idx: number,
  fmt: string | null
): Uint8Array {
  const param = ctx.params[idx];
  if (!param) {
    throw new Error(`renderTemplate: param index ${idx} out of bounds`);
  }
  const offset = paramOffsetAt(ctx.params, paramsData, idx);

  const te = new TextEncoder();

  switch (param.paramType) {
    case ParamType.Address: {
      const bytes = paramsData.subarray(offset, offset + 32);
      return te.encode(bs58.encode(bytes));
    }
    case ParamType.U64: {
      const v = readU64Le(paramsData, offset);
      if (fmt !== null) {
        const decimals = parseDecimalSpec(fmt);
        return te.encode(formatDecimalU64(v, decimals));
      }
      return te.encode(v.toString(10));
    }
    case ParamType.I64: {
      const v = readI64Le(paramsData, offset);
      return te.encode(v.toString(10));
    }
    case ParamType.String: {
      const len = paramsData[offset];
      return paramsData.subarray(offset + 1, offset + 1 + len);
    }
    case ParamType.Bool: {
      return te.encode(paramsData[offset] !== 0 ? "true" : "false");
    }
    case ParamType.U8: {
      return te.encode(paramsData[offset].toString(10));
    }
    case ParamType.U16: {
      const v = paramsData[offset] | (paramsData[offset + 1] << 8);
      return te.encode(v.toString(10));
    }
    case ParamType.U32: {
      const v =
        paramsData[offset] |
        (paramsData[offset + 1] << 8) |
        (paramsData[offset + 2] << 16) |
        (paramsData[offset + 3] << 24);
      // `| 0` yields a signed int; normalise with unsigned shift for the
      // decimal print.
      return te.encode((v >>> 0).toString(10));
    }
    case ParamType.U128: {
      return te.encode(readU128Le(paramsData, offset).toString(10));
    }
    case ParamType.Bytes20: {
      return te.encode("0x" + toHex(paramsData.subarray(offset, offset + 20)));
    }
    case ParamType.Bytes32: {
      return te.encode("0x" + toHex(paramsData.subarray(offset, offset + 32)));
    }
    default: {
      const exhaustive: never = param.paramType;
      throw new Error(`renderTemplate: unknown ParamType ${exhaustive}`);
    }
  }
}

// ── decimal-shift formatter ──────────────────────────────────────────
//
// Mirror of on-chain `push_decimal_u64`:
//   - integer part emitted first
//   - if there's a fractional part, append '.' and `decimals` digits
//     (leading-zero padded), then trim trailing zeros
//   - integer-only values have no decimal point at all
// Examples:
//   (1_000_000_000, 9)  → "1"
//   (   100_000_000, 9) → "0.1"
//   ( 100_000_000_000_000, 18) → "0.0001"

function formatDecimalU64(val: bigint, decimals: number): string {
  if (decimals === 0) return val.toString(10);
  let scale = 1n;
  for (let i = 0; i < decimals; i++) scale *= 10n;
  const intPart = val / scale;
  const fracPart = val % scale;
  let out = intPart.toString(10);
  if (fracPart > 0n) {
    // Leading-zero-pad to `decimals` width.
    let frac = fracPart.toString(10);
    if (frac.length < decimals) frac = "0".repeat(decimals - frac.length) + frac;
    // Trim trailing zeros.
    let end = frac.length;
    while (end > 0 && frac.charCodeAt(end - 1) === 0x30 /* '0' */) end--;
    out += "." + frac.slice(0, end);
  }
  return out;
}

function parseDecimalSpec(spec: string): number {
  if (!spec.startsWith("10^")) {
    throw new Error(
      `renderTemplate: invalid format spec "${spec}"; only "10^<digits>" is supported`
    );
  }
  const rest = spec.slice(3);
  const n = parseDecimalNonNegative(rest, "decimal spec exponent");
  if (n > 19) {
    throw new Error(
      `renderTemplate: decimal shift too large (${n}, max 19)`
    );
  }
  return n;
}

function parseDecimalNonNegative(s: string, field: string): number {
  if (s.length === 0) throw new Error(`renderTemplate: empty ${field}`);
  let n = 0;
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c < 0x30 || c > 0x39) {
      throw new Error(`renderTemplate: non-digit "${s[i]}" in ${field}`);
    }
    n = n * 10 + (c - 0x30);
    if (n > 1e9) throw new Error(`renderTemplate: ${field} too large`);
  }
  return n;
}

// ── small byte-reading helpers (match on-chain little-endian form) ───

function readU64Le(data: Uint8Array, off: number): bigint {
  if (off + 8 > data.length) throw new Error("readU64Le: OOB");
  const dv = new DataView(data.buffer, data.byteOffset + off, 8);
  return dv.getBigUint64(0, /* littleEndian */ true);
}

function readI64Le(data: Uint8Array, off: number): bigint {
  if (off + 8 > data.length) throw new Error("readI64Le: OOB");
  const dv = new DataView(data.buffer, data.byteOffset + off, 8);
  return dv.getBigInt64(0, /* littleEndian */ true);
}

function readU128Le(data: Uint8Array, off: number): bigint {
  if (off + 16 > data.length) throw new Error("readU128Le: OOB");
  const dv = new DataView(data.buffer, data.byteOffset + off, 16);
  const lo = dv.getBigUint64(0, /* littleEndian */ true);
  const hi = dv.getBigUint64(8, /* littleEndian */ true);
  return (hi << 64n) | lo;
}

function decodeAscii(bytes: Uint8Array): string {
  // Templates contain only ASCII inside the `{...}` placeholders; plain
  // ASCII decode is faster than TextDecoder here.
  let s = "";
  for (let i = 0; i < bytes.length; i++) s += String.fromCharCode(bytes[i]);
  return s;
}

function concatBytes(chunks: Uint8Array[]): Uint8Array {
  let total = 0;
  for (const c of chunks) total += c.length;
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

File: crates/clear-msig-intent/src/render.rs
use crate::{
    error::{invalid, IntentSchemaError},
    ParamTypeJson,
};

pub fn render_template(
    template: &str,
    params: &[ParamTypeJson],
    params_data: &[u8],
) -> Result<String, IntentSchemaError> {
    if !template.is_ascii() {
        return Err(invalid("intent template must be ASCII"));
    }
    let mut result = String::new();
    let bytes = template.as_bytes();
    let mut cursor = 0;
    while cursor < bytes.len() {
        if bytes[cursor] != b'{' {
            result.push(bytes[cursor] as char);
            cursor += 1;
            continue;
        }
        let relative_end = bytes[cursor + 1..]
            .iter()
            .position(|byte| *byte == b'}')
            .ok_or_else(|| invalid("intent template has an unmatched '{'"))?;
        let end = cursor + 1 + relative_end;
        let expression = std::str::from_utf8(&bytes[cursor + 1..end])
            .map_err(|_| invalid("intent placeholder must be ASCII"))?;
        let (index_text, format) = expression
            .split_once(':')
            .map_or((expression, None), |(index, format)| (index, Some(format)));
        let index = index_text
            .parse::<usize>()
            .map_err(|_| invalid(format!("invalid template parameter index '{index_text}'")))?;
        result.push_str(&render_param(params, params_data, index, format)?);
        cursor = end + 1;
    }
    Ok(result)
}

fn render_param(
    params: &[ParamTypeJson],
    params_data: &[u8],
    index: usize,
    format: Option<&str>,
) -> Result<String, IntentSchemaError> {
    let param_type = *params
        .get(index)
        .ok_or_else(|| invalid(format!("parameter index {index} is out of bounds")))?;
    if format.is_some() && param_type != ParamTypeJson::U64 {
        return Err(invalid("decimal template formats require u64"));
    }
    let offset = param_offset(params, params_data, index)?;
    match param_type {
        ParamTypeJson::Address => {
            Ok(bs58::encode(param_bytes(params_data, offset, 32, "address")?).into_string())
        }
        ParamTypeJson::U64 => {
            let value = u64::from_le_bytes(
                param_bytes(params_data, offset, 8, "u64")?
                    .try_into()
                    .expect("validated u64 length"),
            );
            match format {
                Some(format) => Ok(format_decimal_u64(value, parse_decimal_spec(format)?)),
                None => Ok(value.to_string()),
            }
        }
        ParamTypeJson::I64 => Ok(i64::from_le_bytes(
            param_bytes(params_data, offset, 8, "i64")?
                .try_into()
                .expect("validated i64 length"),
        )
        .to_string()),
        ParamTypeJson::String => {
            let length = *params_data
                .get(offset)
                .ok_or_else(|| invalid("missing string length"))? as usize;
            let value = param_bytes(params_data, offset + 1, length, "string")?;
            Ok(std::str::from_utf8(value)
                .map_err(|error| invalid(format!("invalid UTF-8 string parameter: {error}")))?
                .to_owned())
        }
        ParamTypeJson::Bool => Ok(if *params_data
            .get(offset)
            .ok_or_else(|| invalid("missing bool parameter"))?
            != 0
        {
            "true"
        } else {
            "false"
        }
        .to_owned()),
        ParamTypeJson::U8 => Ok(params_data
            .get(offset)
            .ok_or_else(|| invalid("missing u8 parameter"))?
            .to_string()),
        ParamTypeJson::U16 => Ok(u16::from_le_bytes(
            param_bytes(params_data, offset, 2, "u16")?
                .try_into()
                .expect("validated u16 length"),
        )
        .to_string()),
        ParamTypeJson::U32 => Ok(u32::from_le_bytes(
            param_bytes(params_data, offset, 4, "u32")?
                .try_into()
                .expect("validated u32 length"),
        )
        .to_string()),
        ParamTypeJson::U128 => Ok(u128::from_le_bytes(
            param_bytes(params_data, offset, 16, "u128")?
                .try_into()
                .expect("validated u128 length"),
        )
        .to_string()),
        ParamTypeJson::Bytes20 => Ok(format!(
            "0x{}",
            encode_hex(param_bytes(params_data, offset, 20, "bytes20")?)
        )),
        ParamTypeJson::Bytes32 => Ok(format!(
            "0x{}",
            encode_hex(param_bytes(params_data, offset, 32, "bytes32")?)
        )),
    }
}

fn parse_decimal_spec(format: &str) -> Result<u8, IntentSchemaError> {
    let decimals = format
        .strip_prefix("10^")
        .ok_or_else(|| invalid(format!("unsupported template format '{format}'")))?
        .parse::<u8>()
        .map_err(|_| invalid(format!("invalid decimal template format '{format}'")))?;
    if decimals > 19 {
        return Err(invalid("decimal template format exceeds 19 digits"));
    }
    Ok(decimals)
}

fn format_decimal_u64(value: u64, decimals: u8) -> String {
    if decimals == 0 {
        return value.to_string();
    }
    let scale = (0..decimals).fold(1u128, |current, _| current * 10);
    let value = value as u128;
    let mut result = (value / scale).to_string();
    let fraction = value % scale;
    if fraction != 0 {
        let mut digits = format!("{:0width$}", fraction, width = decimals as usize);
        while digits.ends_with('0') {
            digits.pop();
        }
        result.push('.');
        result.push_str(&digits);
    }
    result
}

fn param_offset(
    params: &[ParamTypeJson],
    params_data: &[u8],
    target: usize,
) -> Result<usize, IntentSchemaError> {
    let mut offset = 0;
    for param_type in params.iter().take(target) {
        offset += param_size(*param_type, params_data, offset)?;
    }
    Ok(offset)
}

fn param_size(
    param_type: ParamTypeJson,
    params_data: &[u8],
    offset: usize,
) -> Result<usize, IntentSchemaError> {
    match param_type {
        ParamTypeJson::Address | ParamTypeJson::Bytes32 => Ok(32),
        ParamTypeJson::U64 | ParamTypeJson::I64 => Ok(8),
        ParamTypeJson::Bytes20 => Ok(20),
        ParamTypeJson::String => Ok(1 + *params_data
            .get(offset)
            .ok_or_else(|| invalid("missing string length"))?
            as usize),
        ParamTypeJson::Bool | ParamTypeJson::U8 => Ok(1),
        ParamTypeJson::U16 => Ok(2),
        ParamTypeJson::U32 => Ok(4),
        ParamTypeJson::U128 => Ok(16),
    }
}

fn param_bytes<'a>(
    data: &'a [u8],
    offset: usize,
    length: usize,
    label: &str,
) -> Result<&'a [u8], IntentSchemaError> {
    let end = offset
        .checked_add(length)
        .ok_or_else(|| invalid(format!("{label} parameter range overflow")))?;
    data.get(offset..end).ok_or_else(|| {
        invalid(format!(
            "not enough parameter data for {label}: need {offset}..{end}, have {}",
            data.len()
        ))
    })
}

fn encode_hex(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut value = String::with_capacity(bytes.len() * 2);
    for byte in bytes {
        value.push(HEX[(byte >> 4) as usize] as char);
        value.push(HEX[(byte & 0x0f) as usize] as char);
    }
    value
}
